### PR TITLE
RDBC-1035 RDBC-1036 Add aggressive caching context managers and store-level session events

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RDBC-1035
https://issues.hibernatingrhinos.com/issue/RDBC-1036

### Additional description

**RDBC-1035** – Adds aggressive caching support to the Python client. `store.aggressively_cache_for(duration)` and `store.disable_aggressive_caching()` are context managers that enable/disable serving GET responses directly from the local HTTP cache without a server round-trip. Cache options are thread-local, matching the C# `AsyncLocal<T>` semantics. The cache is automatically invalidated when document or index change notifications arrive.

**RDBC-1036** – Adds store-level session event hooks: `add_before_store`, `add_after_save_changes`, `add_before_delete`, and `add_before_query` (with corresponding `remove_*` variants). Handlers registered on the store are automatically wired into every new session, eliminating the need to re-register on each `open_session` call.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low
- [ ] Moderate
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`)
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by poissoncorp** [comment]
please bump PR to run checks

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 1]
```diff
@@ -837,12 +838,14 @@ def delete(self, key_or_entity: Union[str, object], expected_change_vector: Opti
             change_vector = change_vector if self._use_optimistic_concurrency else None
             if self._counters_by_doc_id:
                 self._counters_by_doc_id.pop(key, None)
-            self.defer(
-                DeleteCommandData(
-                    key,
-                    expected_change_vector or change_vector,
-                    expected_change_vector or (document_info.change_vector if document_info is not None else None),
-                )
+            self._pending_key_deletes.append(
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: diverges from C# client which is using defer

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 2]
```diff
@@ -902,6 +905,11 @@ def __store_internal(
                 f"Document id:{key}"
             )
 
+        if any(e["key"] == key for e in self._pending_key_deletes):
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: diverges from C# - should use deleted_entities

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 3]
```diff
@@ -469,6 +469,7 @@ def __init__(self, store: "DocumentStore", key: uuid.UUID, options: SessionOptio
         self._deleted_entities: Union[
             Set[DeletedEntitiesHolder.DeletedEntitiesEnumeratorResult], DeletedEntitiesHolder
         ] = DeletedEntitiesHolder()
+        self._pending_key_deletes: List[Dict] = []
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: what's the origin of this?

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 4]
```diff
@@ -954,11 +962,24 @@ def _store_entity_in_unit_of_work(
         if key is not None:
             self._documents_by_id[key] = document_info
 
+    def __prepare_for_key_deletes(self, result: "InMemoryDocumentSessionOperations.SaveChangesData") -> None:
+        """Fire OnBeforeDelete for string-key deletes and queue their commands.
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: this issue was solved in different PR - this method doesn't exist in C# and doesn't make sense, as we have `__prepare_for_entities_deletion` for that

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 5]
```diff
@@ -1173,7 +1194,11 @@ def has_changes(self) -> bool:
             if self._entity_changed(document, entity.value, None):
                 return True
 
-        return not len(self._deleted_entities) == 0 or not len(self._deferred_commands) == 0
+        return (
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: irrelevant, revert

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 6]
```diff
@@ -1913,6 +1938,7 @@ def __init__(self, session: InMemoryDocumentSessionOperations):
                 self.__documents_by_entity_to_remove: List = []
                 self.__document_infos_to_update: List[Tuple[DocumentInfo, dict]] = []
                 self.__clear_deleted_entities: bool = False
+                self.__clear_pending_key_deletes: bool = False
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: as per previous comments, it's not needed, revert

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 7]
```diff
@@ -1939,8 +1965,14 @@ def clear_session_state_after_successful_save_changes(self):
                 if self.__clear_deleted_entities:
                     self.__session._deleted_entities.clear()
 
+                if self.__clear_pending_key_deletes:
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: as per previous comments, it's not needed, revert

File: ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py [hunk 8]
```diff
@@ -1939,8 +1965,14 @@ def clear_session_state_after_successful_save_changes(self):
...
+                    self.__session._pending_key_deletes.clear()
+
                 self.__session._deferred_commands.clear()
                 self.__session._deferred_commands_map.clear()
 
             def clear_deleted_entities(self) -> None:
                 self.__clear_deleted_entities = True
+
+            def clear_pending_key_deletes(self) -> None:
```
Comment (poissoncorp) [inline ravendb/documents/session/document_session_operations/in_memory_document_session_operations.py]: as per previous comments, it's not needed, revert

File: ravendb/documents/store/definition.py [hunk 9]
```diff
@@ -529,7 +555,122 @@ def initialize(self) -> DocumentStore:
...
+    def aggressively_cache_for(
+        self,
+        cache_duration: datetime.timedelta,
+        database: Optional[str] = None,
+        mode: Optional["AggressiveCacheMode"] = None,
+    ) -> "DocumentStore._AggressiveCacheContext":
+        from ravendb.http.misc import AggressiveCacheOptions, AggressiveCacheMode
+
+        self.assert_initialized()
+        database = self.get_effective_database(database)
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: should follow C# architecture - SetAggressiveCache & FinalizeAggresiveCache

File: ravendb/documents/store/definition.py [hunk 10]
```diff
@@ -529,7 +555,122 @@ def initialize(self) -> DocumentStore:
...
+        if database in self.__aggressive_cache_changes:
+            return
+        # Create the eviction object BEFORE acquiring the lock.
+        # _AggressiveCacheEviction.__init__ calls store.changes() which also
+        # acquires __add_change_lock, so constructing inside the lock would deadlock
+        # (threading.Lock is not reentrant).  If two threads race here both will
+        # build an eviction object, but only one will be stored; the loser is
+        # discarded via close().
+        eviction = DocumentStore._AggressiveCacheEviction(self, database)
+        with self.__add_change_lock:
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py:608]: let's replace the giant comment above with "this lock is here to achieve ConcurrentDict-like behavior" - will increase readability by far

File: ravendb/documents/store/definition.py [hunk 11]
```diff
@@ -529,7 +555,122 @@ def initialize(self) -> DocumentStore:
...
+            if database not in self.__aggressive_cache_changes:
+                self.__aggressive_cache_changes[database] = eviction
+            else:
+                eviction.close()  # another thread won the race
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: missing ensure_connected

File: ravendb/documents/store/definition.py [hunk 12]
```diff
@@ -529,7 +555,122 @@ def initialize(self) -> DocumentStore:
...
+
+    def disable_aggressive_caching(self, database: Optional[str] = None) -> "DocumentStore._AggressiveCacheContext":
+        self.assert_initialized()
+        database = self.get_effective_database(database)
+        request_executor = self.get_request_executor(database)
+        return DocumentStore._AggressiveCacheContext(request_executor, None)
+
+    class _AggressiveCacheContext:
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]:
```suggestion
    class _TemporaryDisabledAggressiveCacheContext:
```

Comment (poissoncorp) [reply ravendb/documents/store/definition.py]: or similar - context manager name needs to indicate the effect on the scope

Comment (poissoncorp) [reply ravendb/documents/store/definition.py]: `_DisableAggressiveCachingContext` ?

**Comment by redknightlois** [comment]
Let me run the cross gap analysis workflow again. This used a much older version (the first draft).

File: ravendb/documents/store/definition.py [hunk 13]
```diff
@@ -529,7 +555,122 @@ def initialize(self) -> DocumentStore:
...
+
+        def __enter__(self):
+            self._old_options = self._request_executor.aggressive_caching
+            self._request_executor.aggressive_caching = self._options
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            self._request_executor.aggressive_caching = self._old_options
+
+    class _AggressiveCacheEviction:
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: Poor naming. Sounds like it's a single eviction, or any other singular context.


```suggestion
    class _AggressiveCacheInvalidator:
    """Subscribes to document/index changes and invalidates the request executor's aggressive cache."""
```

File: ravendb/documents/store/definition.py [hunk 14]
```diff
@@ -379,7 +403,9 @@ def close(self):
         for event in self.__before_close:
             event()
 
-        # todo: evict items from cache based on changes
+        for eviction in list(self.__aggressive_cache_changes.values()):
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: should also change var name here when the class name is improved

File: ravendb/documents/store/definition.py (hunk 13 above)
Comment (redknightlois) [reply ravendb/documents/store/definition.py]: Isnt ` EvictItemsFromCacheBasedOnChanges` which is what we use in C# better?

**Review by poissoncorp — CHANGES_REQUESTED** [review]

File: ravendb/documents/store/definition.py (hunk 13 above)
Comment (poissoncorp) [reply ravendb/documents/store/definition.py]: there's no direct eviction happening inside - I'd stay with invalidation semantic
the "eviction" is implicit

File: ravendb/documents/store/definition.py [hunk 15]
```diff
@@ -529,7 +555,154 @@ def initialize(self) -> DocumentStore:
...
+                )
+            )
+            self._unsubscribers.append(
+                self._changes.for_all_indexes().subscribe_with_observer(
+                    ActionObserver(on_next=on_index_change, on_error=lambda e: _invalidate())
+                )
+            )
+
+        def ensure_connected(self) -> None:
+            # todo: DatabaseChanges auto-connects on __init__; add explicit
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: can be implemented?

File: ravendb/tests/issue_tests/test_RDBC_1035.py [hunk 16]
```diff
@@ -0,0 +1,860 @@
...
+
+    def test_add_and_remove_before_store_roundtrip(self):
+        store = self._make_store_base()
+        handler = lambda e: None
+        store.add_before_store(handler)
+        store.remove_before_store(handler)
+
+    # --- aggressive cache ---
+
+    def test_get_from_cache_returns_empty_release_cache_item_for_non_cacheable_command(self):
```
Comment (poissoncorp) [inline ravendb/tests/issue_tests/test_RDBC_1035.py:59]: can we add happy-path tests proving aggressive cache works? if it's there, correct me please - there's a lot of tests that cover different corner cases

Comment (redknightlois) [reply ravendb/tests/issue_tests/test_RDBC_1035.py:59]: added

File: ravendb/documents/store/definition.py [hunk 17]
```diff
@@ -529,7 +555,152 @@ def initialize(self) -> DocumentStore:
...
+
+    def disable_aggressive_caching(
+        self, database: Optional[str] = None
+    ) -> "DocumentStore._DisableAggressiveCachingContext":
+        self.assert_initialized()
+        database = self.get_effective_database(database)
+        request_executor = self.get_request_executor(database)
+        return DocumentStore._DisableAggressiveCachingContext(request_executor)
+
+    class _AggressiveCacheContext:
```
Comment (poissoncorp) [inline ravendb/documents/store/definition.py]: isn't that duplicated? _DisableAggressiveCachingContext should be used and this class should be removed

**Comment by poissoncorp** [comment]
test_aggressive_cache_serves_second_load_from_cache (leaving the test name here for release note)

File: ravendb/documents/store/definition.py (hunk 17 above)
Comment (poissoncorp) [reply ravendb/documents/store/definition.py]: @redknightlois
